ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -447,24 +449,15 @@ withConnectionManager ConnectionManagerArguments {
             traverse_ cancel (getConnThread connState) )
           state
   where
+    traceConnectionManagerCounters :: StrictTMVar m (ConnectionManagerState peerAddr handle handleError version m) -> m ()
+    traceConnectionManagerCounters stateVar = do
+      mState <- atomically $ readTMVar stateVar >>= traverse readTVar
+      traceWith tracer (TrConnectionManagerCounters (connectionManagerStateToCounters mState))
+
     countConnections :: ConnectionManagerState peerAddr handle handleError version m
                      -> STM m Int
-    countConnections state =
```

Comment:
@coot can you confirm that this code only counts inbound connections? if so then my `numConns` is redundant since there is `inboundConns` defined

Reply:
The only difference I found is that this counter counts `OutboundDupState` and mine does not. Should this state be considered for inbound connections as well?

Reply:
I think that `countConnections` should include `OutboundDupState` as well.  The difference is sublte, and maybe we can avoid including them.   `countConnections` is used to to prune connections so that we never exceeds its limits on inbound connections.  A connection in `OutboundDupState` could actually be accepted (rather than initiated with `connect` call) and later the remote peer stopped using its side, while the local side started using it.  A scenario which explains this:

1. the local node accepts 100 connections
2. all of them negotiated duplex `DataFlow`
3. the local node started using all of them (all are now in `Duplex` state)
4. the remotes node stopped using all of them (they all become `OutboundDupState`)
5. the node accepts more connections, but if we don't count this 100 connections in `OutboundDupState` we can easily be above the limit of connections that we can serve.

If we count `OutboundDupState` as soon as we reach the point 5 we will do the connection pruning (shut down random connections).

If we don't count, pruning will kick in when some number of them will become `DuplexState` (or `InboundState Duplex`). We do pruning when a connection becomes `Duplex`, so this will work but is less conservative.  The `p2p-governor` keeps its own limits how many connections can be in `OutboundDupState Duplex` and `DuplexState` (warm + hot peers), which is around 5 times smaller than the server limits (20 (hot) +50 (warm) against ~300).

Reply:
This makes a lot of sense! Thanks!

Reply:
I would like to see a simple commit which introduces this change (just modifying the `countConnections` functions) and explains it (I would even do that in a separate PR).  Things like that should be clearly visible in the commit history.

Reply:
I made a mistake above, but I think the claim still holds: pruning happens on `DemoteToCold^{Duplex}_{Local} : DuplexState -> InboundState Duplex` transition (not when local side promotes connections).